Use evaluateTransactionFee in calculate-min-fee

newhoggy opened the pull request:
# Changelog

```yaml
- description: |
    Fix calculate-min-fee by switching to use Cardano.Api.evaluateTransactionFee
    instead of the deprecated Cardano.Api.estimateTransactionFee.
    This also deprecates the --mainnet, --testnet-magic, --tx-in-count,
    and --tx-out-count command args, which are no longer necessary.
    They can still be provided, but have no effect.
# uncomment types applicable to the change:
  type:
  - feature        # introduces a new feature
  # - breaking       # the API has changed in a breaking way
  # - compatible     # the API has changed but is non-breaking
  # - optimisation   # measurable performance improvements
  - improvement    # QoL changes e.g. refactoring
  - bugfix         # fixes a defect
  - test           # fixes/modifies tests
  # - maintenance    # not directly related to the code
  # - release        # related to a new release preparation
  # - documentation  # change in code docs, haddocks...
```

# Context

Original PR: https://github.com/IntersectMBO/cardano-cli/pull/534

`Cardano.Api.estimateTransactionFee` (currently used by Cardano CLI `calculate-min-fee`) tends to overestimate transaction fees. Furthermore, inspecting the code, `estimateTransactionFee` appears to only apply the old `a * x + b` fee formula to the transaction size (augmented to take into account the user-provided Shelley and Byron witness counts). This is not how fees should be calculated for later eras.

A deprecation note is currently attached to `Cardano.Api.estimateTransactionFee`, saying that `Cardano.Api.evaluateTransactionFee` should be used instead. Inspecting the code, evaluateTransactionFee calls the Cardano Ledger's era-aware fee calculation function `Cardano.Ledger.Shelley.API.Wallet.evaluateTransactionFee`, which in turn calls the `EraTx` class function `getMinFeeTx` that is instantiated for every Cardano Ledger era.

https://github.com/IntersectMBO/cardano-api/blob/1c10a287093ea1396c4c0ef0c92460f60bab690f/cardano-api/internal/Cardano/Api/Fees.hs#L190

Therefore, cardano-cli should be updated to use Cardano.Api.evaluateTransactionFee in the handler for calculate-min-fee.

https://twitter.com/amw7/status/1740008349905928326

# How to trust this PR

> Highlight important bits of the PR that will make the review faster. If there are commands the reviewer can run to observe the new behavior, describe them.

New golden test to calculate the min fee of a Babbage transaction body with Babbage protocol parameters, comparing against the known minimum fee (below which the transaction gets rejected when submitted to node).

https://github.com/IntersectMBO/cardano-cli/pull/534/files#diff-8d95bc46c5ef4bbab74b92c6b29635f1c3792d1e084d8c7668b0ce75bf9d2b52

# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages
- [x] New tests are added if needed and existing tests are updated. See [Running tests](https://github.com/input-output-hk/cardano-node-wiki/wiki/Running-tests) for more details
- [x] Self-reviewed the diff

<!-- 
### Note on CI ###
If your PR is from a fork, the necessary CI jobs won't trigger automatically for security reasons.
You will need to get someone with write privileges. Please contact IOG node developers to do this
for you. 
-->

smelc commented on cardano-cli/src/Cardano/CLI/EraBased/Options/Common.hs:
```diff
@@ -2231,15 +2244,15 @@ pTxInCount =
     [ Opt.long "tx-in-count"
     , Opt.metavar "NATURAL"
-    , Opt.help "The number of transaction inputs."
+    , Opt.help "DEPRECATED. This argument has no effect."
```
Can you rename the enclosing function to `pTxInCountDeprecated`? Otherwise new callers could be tempted to use this function, where as they should not.

newhoggy replied:
Done!

smelc commented on cardano-cli/src/Cardano/CLI/EraBased/Options/Common.hs:
```diff
@@ -2231,15 +2244,15 @@ pTxInCount =
     [ Opt.long "tx-out-count"
     , Opt.metavar "NATURAL"
-    , Opt.help "The number of transaction outputs."
+    , Opt.help "DEPRECATED. This argument has no effect."
```
Can you rename the enclosing function to `pTxOutCountDeprecated`? Otherwise new callers could be tempted to use this function, where as they should not.

newhoggy replied:
Done!

GeorgeFlerovsky committed:
Use evaluateTransactionFee in calculate-min-fee

Cardano.Api.estimateTransactionFee is deprecated and tends to
over-estimate transaction fees.

There is a deprecation note currently attached to
Cardano.Api.estimateTransactionFee, saying that
Cardano.Api.evaluateTransactionFee should be used instead.

https://github.com/IntersectMBO/cardano-api/blob/1c10a287093ea1396c4c0ef0c92460f60bab690f/cardano-api/internal/Cardano/Api/Fees.hs#L190

Therefore, cardano-cli should be updated to use
Cardano.Api.evaluateTransactionFee in the handler for calculate-min-fee.

The existing golden test for calculate-min-fee was adjusted
and another golden test with a babbage transaction was added.

newhoggy committed:
Reduce code duplication in runTransactionCalculateMinFeeCmd

Jimbo4350 commented on cardano-cli/src/Cardano/CLI/EraBased/Run/Transaction.hs:
```diff
@@ -1009,46 +1012,94 @@ runTransactionCalculateMinFeeCmd :: ()
 
-    UnwitnessedCliFormattedTxBody anyTxBody -> do
-      InAnyShelleyBasedEra sbe txbody <- pure anyTxBody
+    UnwitnessedCliFormattedTxBody (InAnyShelleyBasedEra sbe txbody) -> do
```
We  can get rid `UnwitnessedCliFormattedTxBody`

Jimbo4350 reviewed (commented):
LGTM. We can now use `calcMinFeeTx` for a more accurate fee calculation.
